Use WeakRef / WeakPtr to store non-stack RenderObjects

cdumez opened the pull request:
#### a94d55eae68febe63e553172ddda72561a422461
<pre>
Use WeakRef / WeakPtr to store non-stack RenderObjects
<a href="https://bugs.webkit.org/show_bug.cgi?id=266040">https://bugs.webkit.org/show_bug.cgi?id=266040</a>
<a href="https://rdar.apple.com/118737861">rdar://118737861</a>

Reviewed by Darin Adler.

Use WeakRef / WeakPtr to store non-stack RenderObjects instead of CheckedRef / CheckedPtr.
Crashes generated by CheckedRef / CheckedPtr are not actionable if the smart pointer is
not on the stack.

This tested as performance neutral on the benchmarks we track on macOS and iOS, thanks
to the recent WeakPtr optimizations.

There are still a few PackedCheckedPtr&lt;RenderObject&gt; data members that should be
converted. I&apos;ll look into it in a follow-up given we don&apos;t have a drop-in replacement
at the moment.

* Source/WTF/wtf/WeakRef.h:
* Source/WebCore/editing/SelectionGeometryGatherer.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didDestroyRenderTree):
(WebCore::LocalFrameView::addEmbeddedObjectToUpdate):
(WebCore::LocalFrameView::removeEmbeddedObjectToUpdate):
(WebCore::LocalFrameView::updateEmbeddedObjects):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::addContinuationWithOutline):
(WebCore::RenderBlock::paintContinuationOutlines):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderObject::checkedParent const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isDescendantOf const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderSelection.cpp:
(WebCore::collectSelectionData):
(WebCore::RenderSelection::collectBounds const):
(WebCore::RenderSelection::apply):

Canonical link: <a href="https://commits.webkit.org/271749@main">https://commits.webkit.org/271749@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9ac9153b8ffc8fd3b30f8905a68a5ddfaa7a746

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29509 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8177 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30836 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32052 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26747 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30110 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5469 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26751 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29782 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6834 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25217 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5843 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5974 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26302 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33396 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/25384 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26939 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26722 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32197 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/29756 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5932 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4120 "Found 1 new test failure: fullscreen/full-screen-layer-dump.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29971 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7662 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/26103 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36082 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6470 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7775 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3800 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6455 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3732,24 +3730,12 @@ bool LocalFrameView::updateEmbeddedObjects()
-            weakObject = *embeddedObject;
-        }
-        updateEmbeddedObject(weakObject);
+    unsigned numberOfObjectsToUpdate = m_embeddedObjectsToUpdate->size();
```
This peculiar idiom needs a comment. I see how using the size here prevents an infinite loop locally but no clear understanding of the other reasons it's a good choice and I think it's unconventional enough that we should state the reasons for it.

cdumez committed:
Use WeakRef / WeakPtr to store non-stack RenderObjects
https://bugs.webkit.org/show_bug.cgi?id=266040
rdar://118737861

Reviewed by Darin Adler.

Use WeakRef / WeakPtr to store non-stack RenderObjects instead of CheckedRef / CheckedPtr.
Crashes generated by CheckedRef / CheckedPtr are not actionable if the smart pointer is
not on the stack.

This tested as performance neutral on the benchmarks we track on macOS and iOS, thanks
to the recent WeakPtr optimizations.

There are still a few PackedCheckedPtr<RenderObject> data members that should be
converted. I'll look into it in a follow-up given we don't have a drop-in replacement
at the moment.

* Source/WTF/wtf/WeakRef.h:
* Source/WebCore/editing/SelectionGeometryGatherer.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didDestroyRenderTree):
(WebCore::LocalFrameView::addEmbeddedObjectToUpdate):
(WebCore::LocalFrameView::removeEmbeddedObjectToUpdate):
(WebCore::LocalFrameView::updateEmbeddedObjects):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::addContinuationWithOutline):
(WebCore::RenderBlock::paintContinuationOutlines):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderObject::checkedParent const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isDescendantOf const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderSelection.cpp:
(WebCore::collectSelectionData):
(WebCore::RenderSelection::collectBounds const):
(WebCore::RenderSelection::apply):

Canonical link: https://commits.webkit.org/271749@main
